File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int c = 0;
    int averageOfSubtree(TreeNode* root) {
        fun(root);
        return c;
    }

private:
    pair<int,int> fun(TreeNode* root){
        if(root == NULL){
            return {0,0};
        }
        pair<int,int> left = fun(root -> left);
        pair<int,int> right = fun(root -> right);

        int sum = left.first + right.first + root ->val;
        int n = left.second + right.second + 1;

        if(root -> val == sum / n){
            c++;
        }
        return {sum,n};
    }
};

File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int c = 0;
    public int averageOfSubtree(TreeNode root) {
        fun(root);
        return c;
    }

    private int[] fun(TreeNode root){
        if(root == null){
            return new int[]{0,0};
        }
        int[] left = fun(root.left);
        int[] right = fun(root.right);

        int sum = left[0] + right[0] + root.val;
        int n = left[1] + right[1] + 1;

        if(root.val == sum / n){
            c++;
        }
        return new int[]{sum,n};
    }
}